icu4j/main/classes/core/src/com/ibm/icu/text/RuleBasedBreakIterator.java

```diff
@@ -1436,42 +1436,84 @@ int current() {
-                    // that are safe. If advancing from the safe point moves forwards by less than
-                    // two code points, we need to advance one more time to ensure that the boundary
-                    // is good, including a correct rules status value.
-                    //
-                    fPosition = backupPos;
+        int startOfText = fText.getBeginIndex();
+        int aBoundary = -1;
+        int ruleStatusIndex = 0;
+        boolean retainCache = false;
+        if ((position > fBoundaries[fStartBufIdx] - 15) && position < (fBoundaries[fEndBufIdx] + 15)) {
+            // Requested position is near the existing cache. Retain it.
+            retainCache = true;
+        } else if (position <= startOfText + 20) {
+            // Requested position is near the start of the text. Fill cache from start, skipping
+            // the need to find a safe point.
+            retainCache = false;
+            aBoundary = startOfText;
+        } else {
+            // Requested position is not near the existing cache.
+            // Find a safe point to refill the cache from.
+            int backupPos = handleSafePrevious(position);
+
+            if (fBoundaries[fEndBufIdx] < position && fBoundaries[fEndBufIdx] >= (backupPos - 15)) {
+                // The requested position is beyond the end of the existing cache, but the
+                // reverse rules produced a position near or before the cached region.
+                // Retain the existing cache, and fill from the end of it.
+                retainCache = true;
+            } else if (backupPos < startOfText + 15) {
+                // The safe reverse rules moved us to near the start of text.
+                // Take that (index 0) as the backup boundary, avoiding the complication
+                // (in the following block) of moving forward from the safe point to a known boundary.
+                //
+                // Retain the cache if it begins not too far from the requested position.
+                aBoundary = startOfText;
+                retainCache = (fBoundaries[fStartBufIdx] <= (position + 15));
+            } else {
+                // The safe reverse rules produced a position that is neither near the existing
+                // cache, nor near the start of text.
+                // Advance to the boundary following.
+                // There is a complication: the safe reverse rules identify pairs of code points
+                // that are safe. If advancing from the safe point moves forwards by less than
+                // two code points, we need to advance one more time to ensure that the boundary
+                // is good, including a correct rules status value.
+                //
+                retainCache = false;
+                fPosition = backupPos;
+                aBoundary = handleNext();
+                if (aBoundary == backupPos + 1 ||
+                        (aBoundary == backupPos + 2 &&
```

Comment:
what is the magic number 2 here? Could we add a constant for that and add a note about how we pick that number?

Reply:
backupPos and aBoundary are both string indexes into the input text. The if statement is testing if the difference between them is a single code point. So a difference of 1, or a difference of 2 with the text contents being supplemental pair.

This is Java, so the input is known to be UTF-16. The equivalent function in C++ looks different because it could be either UTF-16 or UTF-8, thanks to UText.